workmail-save-and-update-email/src/app.py

```diff
@@ -0,0 +1,112 @@
+                   { "address" : "recipient2@domain.test" }
+                ]
+            },
+            "sender" : {
+                "address" :  "sender@domain.test"                   # String containing sender email address
+            },
+            "subject" : "Hello From Amazon WorkMail!",              # String containing email subject (Truncated to first 256 chars)"
+            "messageId": "00000000-0000-0000-0000-000000000000",    # String containing message id for retrieval using workmail flow API
+            "invocationId": "00000000000000000000000000000000",     # String containing the id of this lambda invocation. Useful for detecting retries and avoiding duplication
+            "flowDirection": "INBOUND",                             # String indicating direction of email flow. Value is either "INBOUND" or "OUTBOUND"
+            "truncated": false                                      # boolean indicating if any field in message was truncated due to size limitations
+        }
+
+    context: object, required
+    Lambda Context runtime methods and attributes. See https://docs.aws.amazon.com/lambda/latest/dg/python-context-object.html
+
+    Returns
+    -------
+    Amazon WorkMail Sync Lambda Response Format
+    For more information, see https://docs.aws.amazon.com/workmail/latest/adminguide/lambda.html#synchronous-schema
+        return {
+          'actions': [                                              # Required, should contain at least 1 list element
+          {
+            'action' : {                                            # Required
+              'type': 'string',                                     # Required. Can be "BOUNCE", "DROP" or "DEFAULT"
+              'parameters': { <various> }                           # Optional. For bounce, <various> can be {"bounceMessage": "message that goes in bounce mail"}
+            },
+            'recipients': list of strings,                          # Optional. Indicates list of recipients for which this action applies
+            'default': boolean                                      # Optional. Indicates whether this action applies to all recipients
+          }
+        ]}
+
+    """
+
+    logger.info(f"Received event: {event}")
+    email_from = event['envelope']['mailFrom']
+    recipients = event['envelope']['recipients']
+    message_id = event['messageId']
+    logger.info(f"message_id: {message_id}")
+    key = str(uuid.uuid4())
+
+    # Determine if the message is internal
+    if utils.extract_domains([email_from]) == utils.extract_domains(recipients):
+        internal_message = True
+    else:
+        internal_message = False
+        
+    try:
+        # 1. Download email
+        downloaded_email = utils.download_email(message_id)
+        # 2 Save original email
+        if ( not internal_message and os.getenv('SAVE_EXTERNAL_MESSAGES') == "True" ) or ( internal_message and os.getenv('SAVE_INTERNAL_MESSAGES') == "True" ):
```

Comment:
It will make sense to save original message only if it's going to be updated later and therefore we might be able to squash some logic. 

We only require `UPDATE_EXTERNAL_MESSAGES/UPDATE_INTERNAL_MESSAGES` and we can drop `SAVE_EXTERNAL_MESSAGES/SAVE_EXTERNAL_MESSAGES` environment variables.

The updated logic would look something like:

```
update_internal_msg = (utils.get_env_var('UPDATE_INTERNAL_MESSAGES') == 'True')
update_external_msg = (utils.get_env_var('UPDATE_EXTERNAL_MESSAGES') == 'True')
save_and_update_msg = False

if internal_message:
    if update_internal_msg:
        save_and_update_msg = True
else:
    if update_external_msg:
        save_and_update_msg = True

if save_and_update_msg:
    utils.save_email(downloaded_email, key)
    updated_email = utils.update_email(downloaded_email...
    ... rest of the logic...
```

Reply:
I'm trying to think - there may be use cases for saving even if not modifying, but those messages are already in the other bucket (could increase the retention, or trigger another lambda on S3 put object)

Reply:
Indeed. I can't think of any use case as well